### PR TITLE
DocC updates

### PR DESCRIPTION
**Pull request by mbrandonw**
A few pass at taking some of the ideas from @icanzlib's PR #163. More to come soon, but this gets some basics in which should help with our printer documentation.

### CODE REVIEW

File: Sources/Parsing/Builders/OneOfBuilder.swift [hunk 1]
```diff
@@ -1,7 +1,21 @@
-/// A custom parameter attribute that constructs a parser that attempts to run a number of parsers,
-/// one after the other, till one succeeds with an output.
+/// A custom parameter attribute that constructs parsers from closures. The constructed parser
+/// runs each parser in the closure, one after another, till one succeeds with an input.
```
Comment (mbrandonw) [inline Sources/Parsing/Builders/OneOfBuilder.swift:2]: Took some of this wording from the `@ViewBuilder` documentation.

The "custom parameter attribute" is weird to me. Kinda feel like we could say something similar...

File: Sources/Parsing/Documentation.docc/Articles/Parsers/Bool.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+On the other hand, if `Bool.parser()` is used in a context where the input type cannot be inferred,
+then you will get an compiler error:
+
+```swift
+try Parse {
+  Bool.parser()
+  Bool.parser() // 🛑 Ambiguous use of 'parser(of:)'
+}
+.parse("truefalse")
```
Comment (mbrandonw) [inline Sources/Parsing/Documentation.docc/Articles/Parsers/Bool.md:54]: now we have a place to document some of this ambiguity behavior

File: Sources/Parsing/Documentation.docc/Extensions/Parser.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+Some of the most commonly used parsers in the library. Use these parsers with operators in order
+to build complex parsers from simpler pieces.
+
+- <doc:Int>
+- <doc:String>
+- <doc:Bool>
+- <doc:Double>
+- <doc:Float>
+- <doc:CharacterSet>
+- <doc:UUID>
```
Comment (mbrandonw) [inline Sources/Parsing/Documentation.docc/Extensions/Parser.md:32]: now we have a place to document this stuff that isn't currently visible to DocC

File: Sources/Parsing/Builders/OneOfBuilder.swift (hunk 1 above)
Comment (stephencelis) [reply Sources/Parsing/Builders/OneOfBuilder.swift:2]: Do you mean say something other than "custom parameter attribute"?

Comment (stephencelis) [inline Sources/Parsing/Builders/OneOfBuilder.swift:2]: I think previously this meant "succeed by returning an input"

```suggestion
/// runs each parser in the closure, one after another, till one succeeds with an output.
```

But I can see this also being "succeed to parse an input". So depending on which you want to focus on...

File: Sources/Parsing/Documentation.docc/Articles/Parsers/Bool.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
+# Bool
+
+A parser that consumes a Boolean value from the beginning of a string.
+ 
+This parser only recognizes the literal "true" and "false" characters:
```
Comment (stephencelis) [inline Sources/Parsing/Documentation.docc/Articles/Parsers/Bool.md:5]:
```suggestion
This parser only recognizes the literal "true" and "false" sequence of characters:
```

File: Sources/Parsing/Documentation.docc/Articles/Parsers/Bool.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+// error: unexpected input
+//  --> input:1:1
+// 1 | 1 Hello
+//     ^ expected "true" or "false"
+```
+
+The `Bool.parser()` method is overloaded to work on a variety of string representations in order
+to be as efficient as possible, including `Substring`, `UTF8View`, and generally collections of 
+UTF-8 code units (see <doc:StringAbstractions> for more info).
```
Comment (stephencelis) [inline Sources/Parsing/Documentation.docc/Articles/Parsers/Bool.md:30]:
```suggestion
to be as efficient as possible, including `Substring`, `UTF8View`, and more general collections of 
UTF-8 code units (see <doc:StringAbstractions> for more info).
```

File: Sources/Parsing/Documentation.docc/Articles/Parsers/Bool.md (hunk 2 above)
Comment (stephencelis) [reply Sources/Parsing/Documentation.docc/Articles/Parsers/Bool.md:54]: Does that snippet actually fail even with `.parse`? I assume that's enough info, so I guess it's the result builder bug?

File: Sources/Parsing/Documentation.docc/Articles/Parsers/Bool.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+.parse("truefalse")
+```
+
+To fix this you can force one of the boolean parsers to be the `Substring` parser, and then the 
+other will figure it out via type inference:
+
+```swift
+try Parse {
+Bool.parser(of: Substring.self)
+Bool.parser()
```
Comment (stephencelis) [inline Sources/Parsing/Documentation.docc/Articles/Parsers/Bool.md:63]:
```suggestion
  Bool.parser(of: Substring.self)
  Bool.parser()
```

File: Sources/Parsing/Documentation.docc/Articles/Parsers/Bool.md (hunk 5 above)
Comment (stephencelis) [reply Sources/Parsing/Documentation.docc/Articles/Parsers/Bool.md:30]: (Take this or leave it. I think the current way reads OK on seeing it in the other docs, but it did read strange to me initially.)

File: Sources/Parsing/Parser.swift [hunk 7]
```diff
@@ -56,7 +65,63 @@ extension Parser {
...
+  /// ```
+  ///
+  /// > Tip: If your input can have trailing whitespace that you would like to consume and discard
+  /// > you can do so like this:
+  /// > ```swift
+  /// > let output = try Parse {
+  /// >  Int.parser()
+  /// >  ",".utf8
+  /// >  Bool.parser()
+  /// >  Skip { Whitespace() }
```
Comment (stephencelis) [inline Sources/Parsing/Parser.swift:121]:
```suggestion
  /// >   Int.parser()
  /// >   ",".utf8
  /// >   Bool.parser()
  /// >   Skip { Whitespace() }
```

File: Sources/Parsing/Parser.swift [hunk 8]
```diff
@@ -69,7 +134,62 @@ extension Parser {
...
+  /// ```
+  ///
+  /// This method will fail if the parser does not consume the entirety of the input.
+  /// For example:
+  ///
+  /// ```swift
+  /// let output = try Parse {
+  ///  Int.parser()
+  ///  ",".utf8
+  ///  Bool.parser()
```
Comment (stephencelis) [inline Sources/Parsing/Parser.swift:172]:
```suggestion
  ///   Int.parser()
  ///   ",".utf8
  ///   Bool.parser()
```

Comment (stephencelis) [reply Sources/Parsing/Parser.swift:172]: Should this be in a `> Tip:` like above?

**Review by stephencelis — APPROVED** [review]
Looking really good!

File: Sources/Parsing/Parser.swift (hunk 8 above)
Comment (mbrandonw) [reply Sources/Parsing/Parser.swift:172]: I thought we were leaving this part as-is and just making the whitespace thing a tip. You think both should be tips?

File: Sources/Parsing/Documentation.docc/Articles/Parsers/Bool.md (hunk 2 above)
Comment (mbrandonw) [reply Sources/Parsing/Documentation.docc/Articles/Parsers/Bool.md:54]: Oh no it does not, I'll break it out into two lines.

**Pull request merged**